src/analysis.lisp

```diff
@@ -0,0 +1,88 @@
+(in-package :cluck)
+
+(defclass e-analyzer ()
+  ()
+  (:documentation "Simple e-class analyses may be created by creating a subclass of E-CLASS-ANALYZER which implements E-CLASS-ANALYZER-ATOM and E-CLASS-ANALYZER-MERGE. More complex analyses may be implemented by subclassing E-CLASS directly."))
+
+(defclass e-analyzer-cost (e-analyzer)
+  ()
+  (:documentation "An e-class analyzer which can be used as the cost function when determining the best concrete representation of an e-class. Should implement E-CLASS-ANALYZER-COMPARE. A canned implementation of E-ANALYZER-MERGE exists for instances, which simply finds the cheapest e-node according to the analysis value; this implementation must not be overridden."))
+
+(defgeneric e-analyzer-e-node-value (analyzer e-node-car children)
+  (:documentation "Return the analysis value for an e-node. CHILDREN is a list of analysis values of the children. Some elements of CHILDREN may have circular references back to the e-class containing this e-node. In that case, E-ANALYZER-PARENT-VALUE is called to determine a reasonable default value. Note that in some situations, a circular reference will not use the parent value, depending on order of traversal, and instead will just be the value not including this new node."))
+(defgeneric e-analyzer-compare (analyzer v1 v2)
+  (:documentation "Compare two values. Return non-nil iff v1 is strictly less than v2."))
+
+(defgeneric e-analyzer-merge (analyzer v1 v2)
+  (:documentation "Merge the analysis values v1 and v2 into a single value. Ie, if v1 and v2 are values for e-nodes in the same e-class, this method should return the value of the e-class. If v1 and v2 represent costs, that would be the minimum.")
+  (:method ((a e-analyzer-cost) v1 v2)
+    (if (e-analyzer-compare a v1 v2) v1 v2)))
+
+(defgeneric e-analyzer-parent-value (analyzer)
+  (:documentation "See E-ANALYZER-E-NODE-VALUE. Generally speaking, should return some logically maximum/high-cost value for cost analyzers, since a circular reference has infinite size.")
+  (:method (a)
+    'parent))
+
+(defclass tree-size-analyzer (e-analyzer-cost)
+  ()
+  (:documentation "An e-class analyzer that simply counts the smallest representation of each e-class."))
+(defvar +tree-size-analyzer+ (make-instance 'tree-size-analyzer))
```

Comment:
newlines

Reply:
done

(those where like that before I moved them :laughing:)